### ISSUE TITLE
JSX parser should support XML-style comments

### ISSUE THREAD

**Issue by crizzis** [facebook/react#21601]
It would be a nice feature if the JSX parser could recognize XML-style comments (`<!-- -->`), either by ignoring them completely or by passing the comments through to the output HTML. 

Right now, using XML-style comments results in compilation errors.

**Comment by Jarred-Sumner** [comment]
Fun fact: HTML-style comments are valid ECMAScript comments when the host is a web browser (excluding modules)

From the spec: https://tc39.es/ecma262/#sec-html-like-comments (you might have to search "HTML" for the section to appear)

**Comment by kevduc** [comment]
I'm currently looking for a way to add HTML comments to my page (for attributions and other comments potentially useful for whoever reads the HTML source) but as @crizzis said, it's not supported by the JSX parser.

I've found some workarounds here and there, like [Stack Overflow: How to render a HTML comment in React?](https://stackoverflow.com/questions/40015336/how-to-render-a-html-comment-in-react), but these solutions have a lot of boilerplate just for an HTML comment, and I don't think they are easily readable/understandable by someone looking at the code.
So far the most straight forward/concise way of having an HTML comment I've found is with this package: [react-jsx-html-comments](https://www.npmjs.com/package/react-jsx-html-comments)

**TLDR:** I think it would be useful to have HTML comments supported in JSX out of the box to avoid the hassle of finding a workaround/package just to write a comment.

**Comment by JuniorTour** [comment]
I need this feature too.

[Current workaround from stackoverflow](https://stackoverflow.com/questions/40015336/how-to-render-a-html-comment-in-react) has many disadvantages.

We have to:
1. Use DOM operation.
2. Use `dangerouslySetInnerHTML` and get `<div><!-- comment --></div>`, NOT `<!-- comment -->` without `<div>`

**Comment by DRKSLV** [comment]
The thing is, that things like Internet Explorer Comment conditionals can't end within a tag like this `<div><!-- comment --></div>` without causing errors.
Sometimes this is just neccesary to do, espacially in frameworks (e.g next.js) that force you to do EVERYTHING in jsx

**Comment by gaearon** [comment]
We're not actively changing JSX at the moment. But if you'd like to propose a change, https://github.com/facebook/jsx would be the right place.
